runelite-client/src/main/java/net/runelite/client/game/chatbox/ChatboxPanelManager.java

```diff
@@ -0,0 +1,191 @@
+import net.runelite.api.widgets.WidgetInfo;
+import net.runelite.client.callback.ClientThread;
+import net.runelite.client.input.KeyListener;
+import net.runelite.client.input.KeyManager;
+import net.runelite.client.input.MouseListener;
+import net.runelite.client.input.MouseManager;
+import net.runelite.client.input.MouseWheelListener;
+
+@Singleton
+@Slf4j
+public class ChatboxPanelManager
+{
+	private final Client client;
+	private final ClientThread clientThread;
+	private final EventBus eventBus;
+
+	private final KeyManager keyManager;
+	private final MouseManager mouseManager;
+
+	private final Provider<ChatboxTextMenuInput> chatboxTextMenuInputProvider;
+	private final Provider<ChatboxTextInput> chatboxTextInputProvider;
+
+	@Getter
+	private ChatboxInput currentInput = null;
+
+	@Inject
+	private ChatboxPanelManager(EventBus eventBus, Client client, ClientThread clientThread,
+		KeyManager keyManager, MouseManager mouseManager,
+		Provider<ChatboxTextMenuInput> chatboxTextMenuInputProvider, Provider<ChatboxTextInput> chatboxTextInputProvider)
+	{
+		this.client = client;
+		this.clientThread = clientThread;
+		this.eventBus = eventBus;
+
+		this.keyManager = keyManager;
+		this.mouseManager = mouseManager;
+
+		this.chatboxTextMenuInputProvider = chatboxTextMenuInputProvider;
+		this.chatboxTextInputProvider = chatboxTextInputProvider;
+
+		eventBus.register(this);
+	}
+
+	public void close()
+	{
+		clientThread.invokeLater(this::unsafeCloseInput);
+	}
+
+	void unsafeCloseInput()
+	{
+		client.runScript(
+			ScriptID.RESET_CHATBOX_INPUT,
+			0,
+			1
+		);
+	}
+
+	void unsafeOpenInput(ChatboxInput input)
+	{
+		client.runScript(ScriptID.CLEAR_CHATBOX_PANEL);
+
+		eventBus.register(input);
+		if (input instanceof KeyListener)
+		{
+			keyManager.registerKeyListener((KeyListener) input);
+		}
+		if (input instanceof MouseListener)
+		{
+			mouseManager.registerMouseListener((MouseListener) input);
+		}
+		if (input instanceof MouseWheelListener)
+		{
+			mouseManager.registerMouseWheelListener((MouseWheelListener) input);
+		}
+
+		currentInput = input;
+		client.setVar(VarClientInt.INPUT_TYPE, -3);
+		client.getWidget(WidgetInfo.CHATBOX_TITLE).setHidden(true);
+		client.getWidget(WidgetInfo.CHATBOX_FULL_INPUT).setHidden(true);
+
+		Widget c = getContainerWidget();
+		clearDynamicChildren(c);
+		c.setOnDialogAbortListener((JavaScriptCallback) ev -> this.unsafeCloseInput());
+		input.open();
+	}
+
+	public void openInput(ChatboxInput input)
+	{
+		clientThread.invokeLater(() -> unsafeOpenInput(input));
+	}
+
+	public ChatboxTextMenuInput openTextMenuInput(String title)
+	{
+		return chatboxTextMenuInputProvider.get()
+			.title(title);
+	}
+
+	public ChatboxTextInput openTextInput(String prompt)
+	{
+		return chatboxTextInputProvider.get()
+			.prompt(prompt);
+	}
+
+	@Subscribe
+	public void onScriptCallbackEvent(ScriptCallbackEvent ev)
+	{
+		switch (ev.getEventName())
+		{
+			case "resetChatboxInput":
```

Comment:
if this switch is supposed to be extendable in future, maybe add break at the end. otherwise, just if (!ev.getEventName().equals(blahblah)) at the top of the method to reduce unnecessary nesting